### PR TITLE
Fixed #31219 -- Fixed object deletion crash for nested protected related objects.

### PR DESCRIPTION
**Pull request by matthiask**
See https://code.djangoproject.com/ticket/31219#ticket

### CODE REVIEW

File: tests/delete/tests.py [hunk 1]
```diff
@@ -90,6 +90,30 @@ def test_protect_multiple(self):
...
             a.protect.delete()
 
+    def test_protect_chained(self):
+        p = P.objects.create()
+        r = R.objects.create(p=p)
+        B.objects.create(protect=r)
+
+        msg = (
+            "Cannot delete some instances of model 'P' because they are "
+            "referenced through protected foreign keys: 'B.p'."
```
Comment (hramezani) [inline tests/delete/tests.py]: Don't know the error message should be `R.p` or `B.p`?
If it's going to be `R.p` we can fix it by:

```diff
--- a/django/db/models/deletion.py
+++ b/django/db/models/deletion.py
@@ -297,7 +297,7 @@ class Collector:
                         field.remote_field.on_delete(self, field, sub_objs, self.using)
                     except ProtectedError as error:
                         key = "'%s.%s'" % (
-                            error.protected_objects[0].__class__.__name__,
+                            field.model.__name__,
                             field.name,
                         )
                         protected_objects[key] += error.protected_objects
```

@felixxm What do you think?

File: tests/delete/tests.py [hunk 2]
```diff
@@ -90,6 +90,30 @@ def test_protect_multiple(self):
...
+        )
+        with self.assertRaisesMessage(ProtectedError, msg):
+            p.delete()
+
+    def test_restrict_chained(self):
```
Comment (hramezani) [inline tests/delete/tests.py]: I think we don't need this test

File: tests/delete/tests.py (hunk 1 above)
Comment (felixxm) [reply tests/delete/tests.py]: Yes it should be `R.p`, we didn't take into account nested protected relations in the ab3cbd8b9a315911248227208630a020cedca08f (probably my fault). Also casting to list is not necessary anymore after this change.

File: tests/delete/tests.py (hunk 2 above)
Comment (felixxm) [reply tests/delete/tests.py]: Yes we can remove it, it's redundant with `test_restrict_path_cascade_indirect()` and `test_restrict_path_cascade_direct()`, i.e. we have tests for a path "restricted" deletion.

File: tests/delete/tests.py [hunk 3]
```diff
@@ -90,6 +90,30 @@ def test_protect_multiple(self):
         with self.assertRaisesMessage(ProtectedError, msg):
             a.protect.delete()
 
+    def test_protect_chained(self):
```
Comment (felixxm) [inline tests/delete/tests.py]: `test_protect_chained` -> `test_protect_path`

File: tests/delete/tests.py [hunk 4]
```diff
@@ -90,6 +90,30 @@ def test_protect_multiple(self):
...
+        p = P.objects.create()
+        r = R.objects.create(p=p)
+        B.objects.create(protect=r)
```
Comment (felixxm) [inline tests/delete/tests.py]: We can use `create_a()` hook, e.g.
```python
    def test_protect_path(self):
        a = create_a('protect')
        a.protect.p = P.objects.create()
        a.protect.save()
        msg = (
            "Cannot delete some instances of model 'P' because they are "
            "referenced through protected foreign keys: 'R.p'.",
        )   
        with self.assertRaisesMessage(ProtectedError, msg):
            a.protect.p.delete()
```

File: django/db/models/deletion.py [hunk 5]
```diff
@@ -308,7 +308,7 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                     model.__name__,
                     ', '.join(protected_objects),
                 ),
-                chain.from_iterable(protected_objects.values()),
+                list(chain.from_iterable(protected_objects.values())),
```
Comment (felixxm) [inline django/db/models/deletion.py]: We can revert this change and fix collecting protected objects (see [comment](https://github.com/django/django/pull/12394/files#r373228132)).

**Comment by matthiask** [comment]
Thanks @hramezani @felixxm for the review!
